### ISSUE TITLE
[Question] Volume is already exclusively attached to one node and can't be attached to another

### ISSUE THREAD

**Issue by hillbun** [longhorn/longhorn#1985]
I delete statefulset pod that attached to a longhorn PVC.

when POD recreated, it shows errors of:

Multi-Attach error for volume "pvc-0122bbe5-ca54-4e73-87be-5cbe9afa691a" Volume is already exclusively attached to one node and can't be attached to another

how to fix this issue

**Comment by yasker** [comment]
My guess is the volume wasn't completed detached when the statefulset was deleted before recreate. Does it heal it self? The workaround would be scale down the statefulset again, wait for the volume to detach, then scale it back up. If it doesn't detach and you're sure there isn't any other user of the volume, you can detach it from UI. But if that happens, it's a bug.

Which version of Longhorn and Kubernetes you're using?

**Comment by hillbun** [comment]
longhorn v1.0.2
Kubernetes v1.17.9

This error often happened when host resouce CPU, memery high.

How to confirm pvc is complely detach by kubectl command?

**Comment by yasker** [comment]
`kubectl -n longhorn-system get lhv <volume-name>` and then check for `status.state`.

**Comment by hillbun** [comment]
Can not see status

```
 kubectl -n longhorn-system get lhv pvc-caa89e37-db46-4e60-9ae3-a9cb7d4bf99b 
NAME                                       AGE
pvc-caa89e37-db46-4e60-9ae3-a9cb7d4bf99b   34m
```

**Comment by yasker** [comment]
Need to add `-o yaml`, so `kubectl -n longhorn-system get lhv <volume-name> -o yaml`

**Comment by hillbun** [comment]
@yasker 

state: detaching



I have scale statefulset that mounted replicas to 0.
but state keeps detaching for a long time.

**Comment by yasker** [comment]
@hillbun Do you see any error log in the Longhorn manager?

**Comment by alexgorbatchev** [comment]
I'm seeing the same issues every time I update a deployment and a pod gets scheduled on a different node, meaning there's an old pod running on node1 and new pod is scheduled on node2. The `Multi-Attach error for volume` occurs every single time and my deployments never proceed further until I kill the new pod enough time for it get scheduled on the same node where the old pod is currently running.

**Comment by brunomedeirosdedalus** [comment]
@alexgorbatchev  I have seen exactly the same behavior when it is a 'Deployment'; changing it to a 'StatefulSet' seems to solve the issue. Did you manage to find out a proper fix or workaround for your case?

**Comment by Oscar-Campo** [comment]
Same issue here. 

When a node goes down, the new pod is scheduled on another node and it's stuck with the same error above.

Any hints?

**Comment by alexgorbatchev** [comment]
I haven't found a good solution.

**Comment by badihi** [comment]
Same problem here. This is a really serious problem guys. Please look into it, since it is causing us a lot of problems.
@yasker

**Comment by badihi** [comment]
Maybe these logs from longhorn-manager are relevant to this problem:
```
time="2023-02-02T16:29:12Z" level=info msg="NodeUnstageVolume: req: {\"staging_target_path\":\"/var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/83c3dfce43ebefa2c12b2ea9e943ab588102c8772c9dbf97e08fce0220d3b4e9/globalmount\",\"volume_id\":\"pvc-0b6c1249-2084-4784-ae9b-edf76f82ee42\"}"
time="2023-02-02T16:29:12Z" level=debug msg="trying to unmount potential mount point /var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/83c3dfce43ebefa2c12b2ea9e943ab588102c8772c9dbf97e08fce0220d3b4e9/globalmount"
time="2023-02-02T16:29:12Z" level=info msg="no need for unmount not a mount point /var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/83c3dfce43ebefa2c12b2ea9e943ab588102c8772c9dbf97e08fce0220d3b4e9/globalmount"
time="2023-02-02T16:29:12Z" level=debug msg="device /dev/mapper/pvc-0b6c1249-2084-4784-ae9b-edf76f82ee42 is not an active LUKS device: failed to run cryptsetup args: [status pvc-0b6c1249-2084-4784-ae9b-edf76f82ee42] output:  error: exit status 4"
time="2023-02-02T16:29:12Z" level=info msg="NodeUnstageVolume: volume pvc-0b6c1249-2084-4784-ae9b-edf76f82ee42 unmounted from node path /var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/83c3dfce43ebefa2c12b2ea9e943ab588102c8772c9dbf97e08fce0220d3b4e9/globalmount"
time="2023-02-02T16:29:12Z" level=info msg="NodeUnstageVolume: rsp: {}"
```

**Comment by paravibe** [comment]
Same issue. I have to manually delete existing volume and create a new one from backups with different PV name. This is really a big problem.

**Comment by innobead** [comment]
cc @PhanLe1010 

@paravibe Can u provide the support bundle to let us investigate further? Or create another issue to specially describe your case. Thanks.

**Comment by paravibe** [comment]
@innobead what do you mean by support bundle? How can I get it?

**Comment by alexgorbatchev** [comment]
another year passes, still the same problem...

**Comment by PhanLe1010** [comment]
@alexgorbatchev Can you 
1. email us a support bundle (get it from the link on the bottom of Longhorn UI) to investigate?  `longhorn-support-bundle@rancher.com`
2. What is the name of the volume?
3. The yaml and event of workload pod
Thanks

**Comment by alexgorbatchev** [comment]
finally found a way to fix this issue when it happens:

```shell
kubectl get volumeattachment | grep pvc-c7155aa8-282a-461f-b1fb-4260b0687b69
kubectl delete volumeattachment csi-f12fdf2abf305183d7eba30b4012c97a33d33dc0b61612756ec330b0f32901e7
```

**Comment by PhanLe1010** [comment]
Yes, deleting the old VolumeAttachment object would be the workaround but we are interested in finding the root cause if you could help

**Comment by connde** [comment]
Thanks @alexgorbatchev , this was driving me crazy, I was able to use the workaround, even though I manually attached the volume to a node, the attachment was pointing to different node.

@PhanLe1010  I can send whatever is needed if that could help solve the issue

**Comment by PhanLe1010** [comment]
Thanks @connde for willing to help.

Are you able to reproduce the issue in a test cluster? If so, could you provide the reproducing steps and email us a support bundle (get it from the link on the bottom of Longhorn UI) to investigate? `longhorn-support-bundle@rancher.com`

**Comment by connde** [comment]
I'm not sure if I'm able to reproduce it, it happens from time to time, for example now I was upgrading one node, so I drained one node and had all pods on the second node, after the upgrade I proceeded with the upgrade on second node.

What did happened is that I was using almost all CPU resource from the first node and during that a couple of stateful set was unable to attach.

I sent my support bundle, in case that could lead to something, I got two different issues during that operation, I attached both cases with events and pvc id.

**Comment by andrewjameshough** [comment]
> finally found a way to fix this issue when it happens:
> 
> ```shell
> kubectl get volumeattachment | grep pvc-c7155aa8-282a-461f-b1fb-4260b0687b69
> kubectl delete volumeattachment csi-f12fdf2abf305183d7eba30b4012c97a33d33dc0b61612756ec330b0f32901e7
> ```

This answer needs more attention - after a day of digging around this was the solution here. On a hard shutdown the attachment was left hanging and caused no end of issues. This simple fix did the trick and hey presto everything was back.

Much appreciated for this.

**Comment by frenchcomp** [comment]
Hi,
I also have this error, with statefulset. However, the commands below removing the Volumeattachment only resolves the error one time out of three. (and i wait 10 minuts each time)
My cluster runs on microk8s, in the version 1.27.1.
Thanks

**Comment by davidCarlos** [comment]
Hi,

I also have this error, with deployment. I'm testing longhorn with RKE2. When I change the  node affinity (to force the  Pod recreation on a different node) and redeploy the resource, Longhorn cannot mount the volume in the new node.

@PhanLe1010 I think this issue should be reopened.

**Comment by Ansible-man** [comment]
I am also having this issue. Please reopen

**Comment by narayanmurgod** [comment]
I see the same error as well in GKE
`Multi-Attach error for volume <pvcname> Volume is already exclusively attached to one node and can't be attached to another`

**Comment by badihi** [comment]
Problem still persists on v1.7.1.

**Comment by nagylevidotdev** [comment]
still bad on v1.7.2. Is it possible to reopen this issue? Kinda makes rolling updates not so rolling.. with csi deletion added to deployment steps.

**Comment by derekbit** [comment]
@nagylevidotdev Can you provide a support bundle and steps to reproduce?

**Comment by apolik** [comment]
still persists on v1.8.0 during a rolling update. Pod gets stuck in ContainerCreating state

**Comment by derekbit** [comment]
@apolik The ticket is closed. Can you create a new one and attach your support bundle? Thanks.

**Comment by T02200059** [comment]
> > finally found a way to fix this issue when it happens:
> > kubectl get volumeattachment | grep pvc-c7155aa8-282a-461f-b1fb-4260b0687b69
> > kubectl delete volumeattachment csi-f12fdf2abf305183d7eba30b4012c97a33d33dc0b61612756ec330b0f32901e7
> 
> This answer needs more attention - after a day of digging around this was the solution here. On a hard shutdown the attachment was left hanging and caused no end of issues. This simple fix did the trick and hey presto everything was back.
> 
> Much appreciated for this.

👍🏻 这个命令非常有效果，thank u.

**Comment by yangchiu** [comment]
I'm able to reproduce a similar issue based on https://github.com/longhorn/longhorn/issues/1985#issuecomment-854883843 following the steps:

1. Create a deployment:
```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: test-pvc
  namespace: default
  labels:
    test.longhorn.io: e2e
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: longhorn
  resources:
    requests:
      storage: 3221225472  #3Gi
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test-deployment
  labels:
    app: test-deployment
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      app: test-deployment
  template:
    metadata:
      labels:
        app: test-deployment
    spec:
      securityContext:
        runAsUser: 1000
        runAsGroup: 1000
        fsGroup: 1000
      containers:
        - image: alpine/git
          imagePullPolicy: IfNotPresent
          name: sleep
          command: ['/bin/sh', '-c']
          args: ['while true;do date;sleep 5; done']
          volumeMounts:
            - name: pod-data
              mountPath: /data
      volumes:
        - name: pod-data
          persistentVolumeClaim:
            claimName: test-pvc
```

2. Once the workload pod is running, update the deployment yaml to add the node selector to schedule the workload pod to a different node, then apply the change:
```
      nodeSelector:
        kubernetes.io/hostname: ip-10-0-1-151
```

3. Observe the new pod gets stuck in `ContainerCreating` indefinitely:
```
 kubectl get pods -w -owide
NAME                                   READY   STATUS              RESTARTS        AGE   IP           NODE            NOMINATED NODE   READINESS GATES
longhorn-test-minio-54d5c878fb-wrvgd   1/1     Running             21 (4h9m ago)   8h    10.42.1.64   ip-10-0-1-182   <none>           <none>
longhorn-test-nfs-5896ddbf8-q2rtk      1/1     Running             21 (4h9m ago)   8h    10.42.3.27   ip-10-0-1-151   <none>           <none>
test-deployment-64b56df7d9-dvqbm       1/1     Running             0               56m   10.42.1.95   ip-10-0-1-182   <none>           <none>
test-deployment-ff4c8dd-tn8br          0/1     ContainerCreating   0               56m   <none>       ip-10-0-1-151   <none>           <none>
```

With the same `Multi-Attach error` described in this ticket:

```
$ kubectl describe pod test-deployment-ff4c8dd-tn8br
Name:             test-deployment-ff4c8dd-tn8br
Namespace:        default
Priority:         0
Service Account:  default
Node:             ip-10-0-1-151/10.0.1.151
Start Time:       Mon, 23 Jun 2025 16:09:35 +0800
Labels:           app=test-deployment
                  pod-template-hash=ff4c8dd
Annotations:      <none>
Status:           Pending
IP:               
IPs:              <none>
Controlled By:    ReplicaSet/test-deployment-ff4c8dd
Containers:
  sleep:
    Container ID:  
    Image:         alpine/git
    Image ID:      
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/sh
      -c
    Args:
      while true;do date;sleep 5; done
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /data from pod-data (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-twk5m (ro)
Conditions:
  Type                        Status
  PodReadyToStartContainers   False 
  Initialized                 True 
  Ready                       False 
  ContainersReady             False 
  PodScheduled                True 
Volumes:
  pod-data:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  test-pvc
    ReadOnly:   false
  kube-api-access-twk5m:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              kubernetes.io/hostname=ip-10-0-1-151
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason              Age   From                     Message
  ----     ------              ----  ----                     -------
  Normal   Scheduled           56m   default-scheduler        Successfully assigned default/test-deployment-ff4c8dd-tn8br to ip-10-0-1-151
  Warning  FailedAttachVolume  56m   attachdetach-controller  Multi-Attach error for volume "pvc-265f62d8-b422-44b4-8e6c-884e7aea2034" Volume is already used by pod(s) test-deployment-64b56df7d9-dvqbm
```

[supportbundle_6186f146-cc0e-4576-9d41-d0d4397db3ce_2025-06-23T09-07-12Z.zip](https://github.com/user-attachments/files/20861989/supportbundle_6186f146-cc0e-4576-9d41-d0d4397db3ce_2025-06-23T09-07-12Z.zip)

@derekbit Need to you help confirm if it's a valid issue.

**Comment by yangchiu** [comment]
For https://github.com/longhorn/longhorn/issues/1985#issuecomment-1783490726, the behavior looks like the same as https://github.com/longhorn/longhorn/issues/1985#issuecomment-2995607648.

Following the steps to reschedule the workload pod to a different node by setting `nodeAffinity`:

1. Create the deployment:
```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: test-pvc
  namespace: default
  labels:
    test.longhorn.io: e2e
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: longhorn
  resources:
    requests:
      storage: 3221225472  #3Gi
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test-deployment
  labels:
    app: test-deployment
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      app: test-deployment
  template:
    metadata:
      labels:
        app: test-deployment
    spec:
      securityContext:
        runAsUser: 1000
        runAsGroup: 1000
        fsGroup: 1000
      containers:
        - image: alpine/git
          imagePullPolicy: IfNotPresent
          name: sleep
          command: ['/bin/sh', '-c']
          args: ['while true;do date;sleep 5; done']
          volumeMounts:
            - name: pod-data
              mountPath: /data
      volumes:
        - name: pod-data
          persistentVolumeClaim:
            claimName: test-pvc
```

2. Once the workload pod is running, update the deployment yaml to add the nodeAffinity to schedule the workload pod to a different node, then apply the change:
```
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
              - matchExpressions:
                  - key: kubernetes.io/hostname
                    operator: In
                    values:
                      - ip-10-0-1-151
```

3. Observe the new pod gets stuck in ContainerCreating indefinitely:

```
$ kubectl get pods -w -owide
NAME                                     READY   STATUS              RESTARTS       AGE    IP           NODE            NOMINATED NODE   READINESS GATES
e2e-test-deployment-0-57dd45dc5d-6f789   0/1     ContainerCreating   0              87s    <none>       ip-10-0-1-151   <none>           <none>
e2e-test-deployment-0-64fb8b754f-fc7sm   1/1     Running             0              111s   10.42.1.99   ip-10-0-1-182   <none>           <none>
```

```
$ kubectl get deploy e2e-test-deployment-0 -oyaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "2"
  creationTimestamp: "2025-06-24T01:14:45Z"
  generation: 2
  labels:
    app: e2e-test-deployment-0
    test.longhorn.io: e2e
  name: e2e-test-deployment-0
  namespace: default
  resourceVersion: "178932"
  uid: 9372b3eb-c0f4-4804-b035-3ba844daf096
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: e2e-test-deployment-0
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: e2e-test-deployment-0
        test.longhorn.io: e2e
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: kubernetes.io/hostname
                operator: In
                values:
                - ip-10-0-1-151
      containers:
      - args:
        - while true;do date;sleep 5; done
        command:
        - /bin/sh
        - -c
        image: alpine/git
        imagePullPolicy: IfNotPresent
        name: sleep
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /data
          name: pod-data
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext:
        fsGroup: 1000
        runAsGroup: 1000
        runAsUser: 1000
      terminationGracePeriodSeconds: 30
      volumes:
      - name: pod-data
        persistentVolumeClaim:
          claimName: e2e-test-claim-0
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2025-06-24T01:14:58Z"
    lastUpdateTime: "2025-06-24T01:14:58Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2025-06-24T01:14:45Z"
    lastUpdateTime: "2025-06-24T01:15:09Z"
    message: ReplicaSet "e2e-test-deployment-0-57dd45dc5d" is progressing.
    reason: ReplicaSetUpdated
    status: "True"
    type: Progressing
  observedGeneration: 2
  readyReplicas: 1
  replicas: 2
  unavailableReplicas: 1
  updatedReplicas: 1
```

```
$ kubectl rollout status deployment/e2e-test-deployment-0
Waiting for deployment "e2e-test-deployment-0" rollout to finish: 1 old replicas are pending termination...
```

```
$ kubectl describe pods e2e-test-deployment-0-675b86dbbd-jm8cz
......
  Warning  FailedAttachVolume  45s   attachdetach-controller  Multi-Attach error for volume "pvc-8033cfea-e19f-46f3-a6ab-c9ad4128e323" Volume is already used by pod(s) e2e-test-deployment-0-8d549c5c4-rl9p7

```

**Comment by yangchiu** [comment]
Manually deleting the `volumeattachment`, as suggested in https://github.com/longhorn/longhorn/issues/1985#issuecomment-1560158095, sometimes helps solve the issues in https://github.com/longhorn/longhorn/issues/1985#issuecomment-2995607648 and https://github.com/longhorn/longhorn/issues/1985#issuecomment-2998450426.

If the newly created `volumeattachment` happens to be created on the new node, the old workload can be automatically terminated and the roll out proceeds. 

But if it still be recreated on the old node, manually deleting the `volumeattachment` may need to be repeated multiple times until it's created on the new node. So it's understandable that some users found the workaround didn’t work as expected: https://github.com/longhorn/longhorn/issues/1985#issuecomment-1567736808

**Comment by yangchiu** [comment]
@derekbit Feel free to reopen this ticket if you think there's an issue.

**Comment by goldyfruit** [comment]
Just had the issue with Longhorn 1.8.2 (via Harvester)
